examples_tests/61.BlitFilterTest/blit/formats/r11f_g11f_b10f.glsl

```diff
@@ -0,0 +1,59 @@
+#ifndef _NBL_GLSL_BLIT_R11FG11FB10F_INCLUDED_
+#define _NBL_GLSL_BLIT_R11FG11FB10F_INCLUDED_
+
+uint nbl_glsl_encode_ufloat(in float _f32, in uint mantissaMask, in uint mantissaBits)
+{
+	const uint f32 = floatBitsToUint(_f32);
+
+	if ((f32 & 0x80000000u) != 0u) // negative numbers converts to 0 (represented by all zeroes)
+		return 0;
```

Comment:
1. I don't understand what `maxSinglePrecisionVal` is supposed to be? Is it $3.4028237 \times 10^{38}$ (`FLT_MAX`)?

2. What function would have the return value `expMask|mantissaMask`, because it seems like that would always evalutate to `mantissaMask` and I'm not sure how that is useful?

3. Can't we just clamp the input value to min and max possible values of r11fg11fb10f at the start of the encoding function, [like we do for the shared exponent format](https://github.com/Devsh-Graphics-Programming/Nabla/blob/5588c69ac714734ca50b5d3c1409fed88b258bff/include/nbl/builtin/glsl/format/encode.glsl#L25), then perhaps we would not have to handle overflow and underflow cases separately at all?

Reply:
> 1. I don't understand what maxSinglePrecisionVal is supposed to be? Is it  (FLT_MAX)?

The maximum value of the small float format you're encoding to, but as a `uint` of the 32-bit representation of the number. You can compute it knowing the mantissa mask, whether there is an implicit leading `1.`, the maximum exponent which does not result in INF/NaN and the exponent bias (both you can work out from the number of exponent bits).

Some formats (only RGB9E5 really) don't have an implicit `1.` leading (because it would be impossible to set one of the channel values to 0 with a shared exponent), so the way the max value is computed is slightly different. I'm also unsure whether RGB9E5 supports Inf and NaN.

> 2. What function would have the return value expMask|mantissaMask, because it seems like that would always evalutate to mantissaMask and I'm not sure how that is useful?

You enter the `if()` statement only if the value was encode-able, otherwise you return INF or NaN.

I guess I should have written
```cpp
if (_f32<uintBitsToFloat(maxSinglePrecisionVal)) // NaN always compares false, so this catches both out of bounds and NaN and turns them to INF
{
    if (_f32<0.f)
        return 0.f; 
    // extract mantissa and exponent, truncate and re-bias
    ...
    return encoded_value;
}
return expMask|(isnan(_f32) ? mantissaMask:0u);
```

> 3. Can't we just clamp the input value to min and max possible values of r11fg11fb10f at the start of the encoding function, [like we do for the shared exponent format](https://github.com/Devsh-Graphics-Programming/Nabla/blob/5588c69ac714734ca50b5d3c1409fed88b258bff/include/nbl/builtin/glsl/format/encode.glsl#L25), then perhaps we would not have to handle overflow and underflow cases separately at all?

Clamping won't work if the input is a NaN, it always compares false and messes everything up.

Reply:
Okay but I think it should be:
```GLSL
// Since 11bit and 10bit float only differs in the number of bits in mantissa their min and max values would be similar
// This page https://www.khronos.org/opengl/wiki/Small_Float_Formats#R11F_G11F_B10F, uses the same for both.
const uint minSinglePrecisionVal =  floatBitsToUint(MIN_R11G11B10);
const uint maxSinglePrecisionVal = floatBitsToUInt(MAX_R11G11B10);
if (_f32<uintBitsToFloat(maxSinglePrecisionVal)) // NaN always compares false, so this catches both out of bounds and NaN and turns them to INF
{
    if (_f32<uintBitsToFloat(minSinglePrecisionVal))
        return 0.f; 
    // extract mantissa and exponent, truncate and re-bias
    ...
    return encoded_value;
}
return expMask|(isnan(_f32) ? mantissaMask:0u);
```

Reply:
Resolved.

Reply:
> Okay but I think it should be:
> 
> ```glsl
> // Since 11bit and 10bit float only differ in the number of bits in mantissa their min and max values would be similar
> // This page https://www.khronos.org/opengl/wiki/Small_Float_Formats#R11F_G11F_B10F, uses exactly the same values for both 10 and 11 bit floats.
> const uint minSinglePrecisionVal = floatBitsToUint(MIN_R11G11B10);
> const uint maxSinglePrecisionVal = floatBitsToUInt(MAX_R11G11B10);
> if (_f32<uintBitsToFloat(maxSinglePrecisionVal)) // NaN always compares false, so this catches both out of bounds and NaN and turns them to INF
> {
>     if (_f32<uintBitsToFloat(minSinglePrecisionVal))
>         return 0.f; 
>     // extract mantissa and exponent, truncate and re-bias
>     ...
>     return encoded_value;
> }
> return expMask|(isnan(_f32) ? mantissaMask:0u);
> ```

if you have one more mantissa bit available with the same exponent, then it means your max value is `Exponent*0.5/2^lastBit` higher

Reply:
see my spec comments.